diff --git a/ultralytics/nn/modules/conv.py b/ultralytics/nn/modules/conv.py
index 6c15e1d6..40b3651d 100644
--- a/ultralytics/nn/modules/conv.py
+++ b/ultralytics/nn/modules/conv.py
@@ -37,7 +37,7 @@ def autopad(k, p=None, d=1):  # kernel, padding, dilation
 class Conv(nn.Module):
     """Standard convolution with args(ch_in, ch_out, kernel, stride, padding, groups, dilation, activation)."""

-    default_act = nn.SiLU()  # default activation
+    default_act = nn.Hardswish()  # default activation

     def __init__(self, c1, c2, k=1, s=1, p=None, g=1, d=1, act=True):
         """Initialize Conv layer with given arguments including activation."""

